### PR TITLE
feature: Add simulator module with mcm simulator implementation

### PR DESCRIPTION
**Pull request by rmshaffer**
*Issue #, if available:*
Merging this PR unblocks the following task: [Publish autoqasm to PyPI](https://github.com/orgs/amazon-braket/projects/2/views/1?pane=issue&itemId=62905956)

*Description of changes:*
Ports the contents of https://github.com/amazon-braket/amazon-braket-default-simulator-python/tree/mcm-sim into a new experimental simulator module in AutoQASM as a separate simulator implementation.

The simulator can be used by instantiating the Braket LocalSimulator object with the "autoqasm" backend: 

```
import autoqasm as aq
from braket.devices.local_simulator import LocalSimulator

@aq.main
def bell_state():
    h(0)
    cnot(0, 1)
    return measure([0, 1])

device = LocalSimulator("autoqasm")
result = device.run(bell_state, shots=100).result()
```

*Testing done:*
Tests added to cover 100% of new code.

## To do before merging

- [x] Update docs and example notebooks to show how to use `McmSimulator` (and so that `tox -e notebooks` passes).
- [x] Add tests to reach 100% code coverage in the new code.
- [x] Finish all docstrings where there are `_summary_` and `_description_` placeholders.
- [x] Wait for https://github.com/amazon-braket/amazon-braket-sdk-python/pull/976 to be merged, then update this PR to depend on a released version of amazon-braket-sdk rather than a commit hash.

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/amazon-braket/autoqasm/blob/main/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/amazon-braket/autoqasm/blob/main/CONTRIBUTING.md#commit-your-change)
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/amazon-braket/autoqasm/blob/main/README.md) and [API docs](https://github.com/amazon-braket/autoqasm/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -25,11 +25,8 @@
     package_dir={"": "src"},
     install_requires=[
         # Pin the latest commit of mcm-sim branch of amazon-braket/amazon-braket-sdk-python.git
-        # and amazon-braket/amazon-braket-default-simulator-python.git to get the version of the
-        # simulator that supports the mcm=True argument for Monte Carlo simulation of mid-circuit
-        # measurement, which AutoQASM requires.
-        "amazon-braket-sdk @ git+https://github.com/amazon-braket/amazon-braket-sdk-python.git@ff73de68cf6ac2d0a921e8fe62693e5b9ae2e321#egg=amazon-braket-sdk",  # noqa E501
-        "amazon-braket-default-simulator @ git+https://github.com/amazon-braket/amazon-braket-default-simulator-python.git@ab068c860963c29842d7649c741f88da669597eb#egg=amazon-braket-default-simulator",  # noqa E501
+        "amazon-braket-sdk @ git+https://github.com/amazon-braket/amazon-braket-sdk-python.git@677d8107225098e32a127d3c7a4a8767d63f9d7a#egg=amazon-braket-sdk",  # noqa E501
```
Comment (speller26) [inline setup.py:28]: Why do we need a pinned SDK?

File: src/autoqasm/simulator/program_context.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+    SymbolLiteral,
+)
+from braket.default_simulator.openqasm.program_context import ProgramContext, Table
+from braket.default_simulator.operation import GateOperation
+from sympy import Integer
+
+from autoqasm.simulator.conversion import convert_to_output
+
+
+class QubitTable(Table):
```
Comment (speller26) [inline src/autoqasm/simulator/program_context.py:42]: Any reason not to just add this to the main branch of the default simulator's `QubitTable`?

File: setup.py (hunk 1 above)
Comment (rmshaffer) [reply setup.py:28]: We won't, as soon as the new BDK release comes out tomorrow - will update this to a version requirement then.

**Review by yitchen-tim — DISMISSED** [review]

File: src/autoqasm/simulator/program_context.py (hunk 2 above)
Comment (rmshaffer) [reply src/autoqasm/simulator/program_context.py:42]: No, I don't see any reason why we couldn't do that. I have opened issue https://github.com/amazon-braket/autoqasm/issues/20 to track this.

File: src/autoqasm/simulator/__init__.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+simulate AutoQASM programs at the full OpenQASM 3.0 scope of functionality,
+including programs which contain mid-circuit measurement and classical control flow.
+
+The recommended usage of this simulator is by instantiating the Braket LocalSimulator
+object with the "autoqasm" backend:
+
+.. code-block:: python
+
+    import autoqasm as aq
+    from braket.devices.local_simulator import LocalSimulator
```
[the remark this answers is not in the document]
Comment (rmshaffer) [reply src/autoqasm/simulator/__init__.py]: Thanks, fixed this everywhere.

File: test/unit_tests/autoqasm/test_simulator.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
```
[the remark this answers is not in the document]
Comment (rmshaffer) [reply test/unit_tests/autoqasm/test_simulator.py:13]: Good point, I'll add a test for this.

**Review by yitchen-tim — APPROVED** [review]

**Pull request merged**